twitter-as.js

```diff
@@ -0,0 +1,226 @@
+        if (tuser != null) {
+            if (!tuser.protected) {
+                return constructTimelineRoom(tuser, aliasLocalpart);
+            }
+        }
+    });
+}
+
+function eventQuery(request, context) {
+    var event = request.getData();
+    console.log(event.type);
+    if (event.type == "m.room.member") {
+        if (event.membership == "invite" && event.state_key.startsWith("@twitter_")) { //We should prolly use a regex
+            var intent = bridge.getIntent(event.state_key);
+            intent.join(event.room_id);
+
+            //Set the avatar based on the 'owners' avatar.
+            intent.getClient().getProfileInfo(event.state_key, 'avatar_url').then((url) => {
+                console.log("Set Room Avatar:", url);
+                intent.sendStateEvent(event.room_id, "m.room.avatar", "", {
+                    "url": url.avatar_url
+                });
+            });
+
+            if (context.rooms.remote != null) {
+                Twitter.add_timeline(event.state_key, context.rooms.matrix, context.rooms.remote);
+            } else {
+                console.log("Couldn't find the remote room for this timeline.");
+            }
+        }
+    }
+
+    //console.log("Request:",request);
+    //console.log("Context:",context);
+    return; // we will handle incoming matrix requests later
+}
+
+/*
+  This will create a stream room for one users timeline.
+  Users who are not authenticated via OAuth will recieve the default power of 0
+  Users who are authenticated via Oauth will recieve a power level of 10
+  The owner of this stream will recieve a 75
+  The bot will have 100
+*/
+function constructTimelineRoom(user, aliasLocalpart) {
+    var botID = bridge.getBot().getUserId();
+
+    var roomOwner = "@twitter_" + user.id_str + ":" + bridge.opts.domain;
+    var users = {};
+    users[botID] = 100;
+    users[roomOwner] = 75;
+    var powers = roomPowers(users);
+    var remote = new RemoteRoom("timeline_" + user.id_str);
+    remote.set("twitter_type", "timeline");
+    remote.set("twitter_user", roomOwner);
+    opts = {
+        visibility: "public",
+        room_alias_name: aliasLocalpart,
+        name: "[Twitter] " + user.name,
+        topic: user.description,
+        invite: [roomOwner],
+        initial_state: [
+            powers, {
+                "type": "m.room.join_rules",
+                "content": {
+                    "join_rule": "public"
+                },
+                "state_key": ""
+            }, {
+                "type": "org.twitter.data",
+                "content": user,
+                "state_key": ""
+            }
+        ]
+    };
+    return {
+        creationOpts: opts,
+        remote: remote
+    };
+}
+
+//WIP
+function uploadImageFromUrl(bridge, url, id = null, name = null) {
+    var contenttype;
+    return new Promise((resolve, reject) => {
+        https.get((url), (res) => {
+            contenttype = res.headers["content-type"];
```

Comment:
Even though this function is called `uploadImageFromUrl`, it is a lie. There's nothing specific here to images. If you want it to be generic content that's fine, rename the function. If you only want it to be images, you should do at least *some* due diligence and check the `contenttype` for `image/`. It would also be nice to make it clear what the ultimate return is (a Promise which resolves to a `mxc://` content URI).